src/mpi/romio/adio/common/ad_read_coll.c

```diff
@@ -51,6 +51,25 @@ void ADIOI_Fill_user_buffer(ADIO_File fd, void *buf, ADIOI_Flatlist_node
 				   ADIO_Offset *fd_end,
 				   MPI_Aint buftype_extent);
 
+#define LUSTRE_RD_LOCK_AHEAD 1
```

Comment:
Unpleasant layering violation.  You could 
* add "lockahead" function to the ADIOI_Fns function pointers and implement it for every file system (a bunch of work for you, yes), 
* Better idea: can you have `ADIOI_LUSTRE_IOContig()` manage lustre locks?  You still have access to fd->hints->cb_nodes .  Also minimizes changes you'd make to lustre write_contig, too, right?